hanging_threads.py

```diff
@@ -91,35 +99,65 @@ def monitor(seconds_frozen, tests_per_second):
     of test_per_second parameter
     """
     current_thread = threading.current_thread()
-    self = get_ident()
-    old_threads = {}
+    hanging_threads = set()
+    old_threads = {}  # Threads found on previous iteration.
+
     while not current_thread.is_stopped():
+        new_threads = get_current_frames()
+
+        # Report died threads.
+        for thread_id in old_threads.keys():
+            if thread_id not in new_threads and thread_id in hanging_threads:
+                log_died_thread(thread_id)
+
+        # Process live threads.
         time.sleep(1. / tests_per_second)
         now = time.time()
         then = now - seconds_frozen
-        frames = sys._current_frames()
-        new_threads = {}
-        for frame_id, frame in frames.items():
-            new_threads[frame_id] = thread2list(frame)
-        for thread_id, frame_list in new_threads.items():
-            if thread_id == self: continue
-            if thread_id not in old_threads or \
-               frame_list != old_threads[thread_id][0]:
-                new_threads[thread_id] = (frame_list, now)
-            elif old_threads[thread_id][1] < then:
-                print_frame_list(frame_list, thread_id)
+        for thread_id, thread_data in new_threads.items():
+            # Don't report the monitor thread.
+            if thread_id == current_thread.ident:
+                continue
+            frame = thread_data['frame']
+            # If thread is new or it's stack is changed then update time.
+            if (thread_id not in old_threads or
+                    frame != old_threads[thread_id]['frame']):
+                thread_data['time'] = now
+                # If the thread was hanging then report awaked thread.
+                if thread_id in hanging_threads:
+                    hanging_threads.remove(thread_id)
+                    log_awaked_thread(thread_id)
             else:
-                new_threads[thread_id] = old_threads[thread_id]
+                # If stack is not changed then keep old time.
+                last_change_time = old_threads[thread_id]['time']
+                thread_data['time'] = last_change_time
+                # Check if this is a new hanging thread.
+                if (thread_id not in hanging_threads and
+                        last_change_time < then):
+                    # Gotcha!
+                    hanging_threads.add(thread_id)
+                    # Report the hanged thread.
+                    log_hanged_thread(thread_id, frame)
         old_threads = new_threads
```

Comment:
I have to say that the `monitor()` function becomes too large. However I have not found simple elegant way to divide it into small chunks. I think it should be rewritten in FP or OOP style. But this is probably a job for another task. For now I just want to make it usable for me and others to let us all fix the hanging threads problem in our apps :)